### PR TITLE
Commit to enable AWS multi avail zone

### PR DESCRIPTION
**Pull request by mazzystr**
Commit to enable AWS multi avail zone!

Changes -
- roles/openshift_aws/defaults/main.yml - Additional subnets in openshift_aws_vpc are unused. We should not deploy them! We will now comment them by default. Additional elements will trigger subnets, ELB subnets, and ASG subnets. VPC items are relocated near the top of the infrastructure component list.
- roles/openshift_aws/tasks/elb_single.yml - Ensure subnets listed in openshift_aws_vpc dict are deployed.
- roles/openshift_aws/tasks/scale_group.yml - Ensure subnets listed in openshift_aws_vpc dict are deployed.
- roles/openshift_aws/tasks/vpc.yml - Adding a quick check to ensure the number of az's in openshift_aws_vpc are available in the accounts region.
- roles/openshift_aws/tasks/vpc_and_subnet_id.yml - Add with_items loops to deploy subnets and attempt to retain "default subnet" behavior.

Sorry about closing pull/9366.  I had to move this to a branch to avoid conflicts.

### CODE REVIEW

**Comment by mazzystr** [comment]
### **BEFORE**

> \# oc get nodes --show-labels=true
> NAME                            STATUS    ROLES     AGE       VERSION           LABELS
> ip-172-31-28-192.ec2.internal   Ready     master    50m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kubernetes.io/hostname=ip-172-31-28-192.ec2.internal,node-role.kubernetes.io/master=true
> ip-172-31-28-236.ec2.internal   Ready     compute   44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kubernetes.io/hostname=ip-172-31-28-236.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-45-159.ec2.internal   Ready     master    50m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1e,kubernetes.io/hostname=ip-172-31-45-159.ec2.internal,node-role.kubernetes.io/master=true
> ip-172-31-45-254.ec2.internal   Ready     infra     44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1e,kubernetes.io/hostname=ip-172-31-45-254.ec2.internal,node-role.kubernetes.io/infra=true
> ip-172-31-53-121.ec2.internal   Ready     compute   44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-53-121.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-54-212.ec2.internal   Ready     master    50m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-54-212.ec2.internal,node-role.kubernetes.io/master=true
> ip-172-31-58-76.ec2.internal    Ready     compute   8m        v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-58-76.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-60-50.ec2.internal    Ready     infra     44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-60-50.ec2.internal,node-role.kubernetes.io/infra=true


### **AFTER infra and compute scale up**

> \# oc get nodes --show-labels=true
> NAME                            STATUS    ROLES     AGE       VERSION           LABELS
> ip-172-31-19-255.ec2.internal   Ready     compute   8m        v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kubernetes.io/hostname=ip-172-31-19-255.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-28-160.ec2.internal   Ready     infra     18m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kubernetes.io/hostname=ip-172-31-28-160.ec2.internal,node-role.kubernetes.io/infra=true
> ip-172-31-28-192.ec2.internal   Ready     master    50m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kubernetes.io/hostname=ip-172-31-28-192.ec2.internal,node-role.kubernetes.io/master=true
> ip-172-31-28-236.ec2.internal   Ready     compute   44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kubernetes.io/hostname=ip-172-31-28-236.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-40-72.ec2.internal    Ready     compute   44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1e,kubernetes.io/hostname=ip-172-31-40-72.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-45-159.ec2.internal   Ready     master    50m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1e,kubernetes.io/hostname=ip-172-31-45-159.ec2.internal,node-role.kubernetes.io/master=true
> ip-172-31-45-173.ec2.internal   Ready     compute   8m        v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1e,kubernetes.io/hostname=ip-172-31-45-173.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-45-254.ec2.internal   Ready     infra     44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1e,kubernetes.io/hostname=ip-172-31-45-254.ec2.internal,node-role.kubernetes.io/infra=true
> ip-172-31-53-121.ec2.internal   Ready     compute   44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-53-121.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-54-212.ec2.internal   Ready     master    50m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-54-212.ec2.internal,node-role.kubernetes.io/master=true
> ip-172-31-58-76.ec2.internal    Ready     compute   8m        v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-58-76.ec2.internal,node-role.kubernetes.io/compute=true
> ip-172-31-60-50.ec2.internal    Ready     infra     44m       v1.10.0+b81c8f8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m4.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1c,kubernetes.io/hostname=ip-172-31-60-50.ec2.internal,node-role.kubernetes.io/infra=true

File: roles/openshift_aws/tasks/vpc_and_subnet_id.yml [hunk 1]
```diff
@@ -15,8 +15,14 @@
...
-      "availability-zone": "{{ openshift_aws_subnet_az }}"
       vpc-id: "{{ vpcout.vpcs[0].id }}"
+      "tag:Name": "{{ item }}"
+  with_items: ['{%- if (openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute="az") | list | length) == 1 -%}
+      {{ openshift_aws_subnet_az }}
+    {%- elif (openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute="az") | list | length) > 1 -%}
+      {{ openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute="az") | list }}
+    {%- else -%}
+    {%- endif -%}']
   register: subnetout
```
Comment (mazzystr) [inline roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: screen shot 2018-08-01 at 1 32 41 pm

Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: @mwoodson @sdodson @joelddiaz

Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: Simply using ...
```{{ openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute="az") | list }}```
here would be awesome.

That would abandon the default_az subkey in openshift_aws_vpc, key openshift_aws_subnet_az, and filter get_default_az.  I would love to get rid of all those bits as long as we all agree.

@abutcher in regards to cluster-operator, what do you think about removing default_az and using element 0 as default az?

Comment (abutcher) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: I like the explicit default az attribute but I don't feel very strongly about it.

Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: If we remove default az will CO be affected at all?

Comment (joelddiaz) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: what do you mean by "remove default az"?

Comment (abutcher) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: He means remove the `default_az` field we set here: https://github.com/openshift/cluster-operator/blob/c2f5837cd1ac8b8ed752d42eb5d3dca42ababec4/pkg/ansible/generate.go#L44

I believe our default is driven by an arg here https://github.com/openshift/cluster-operator/blob/489a78eb1bf85d43964d3de61654aee4e71e6f08/cmd/aws-machine-controller/machine-controller.go#L52 so I don't think removing `default_az` will impact us.

Comment (abutcher) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: We use the `default_az` in the inventory when provisioning the ELB since we rely on openshift-ansible for that. So that's the only place it may impact us.

Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: re deploying ELBs ... it would just use element 0 for single az deployments

File: roles/openshift_aws/tasks/vpc.yml [hunk 2]
```diff
@@ -47,6 +62,6 @@
     tags:
       Name: "{{ openshift_aws_vpc_name }}"
     routes:
-    - dest: 0.0.0.0/0
-      gateway_id: igw
+      - dest: 0.0.0.0/0
```
Comment (michaelgugino) [inline roles/openshift_aws/tasks/vpc.yml]: undo this change.

File: roles/openshift_aws/tasks/vpc.yml [hunk 3]
```diff
@@ -1,4 +1,19 @@
...
+  register: azs
+
+- debug:
+    msg:
+      - "AWS VPC does NOT contain
+      {{ openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute='az') | list | length }}
+      availability zone(s)."
+      - "Please pick another region!"
+  when:
+    - ( azs.availability_zones | length | int ) < ( openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute='az') | list | length )
```
Comment (michaelgugino) [inline roles/openshift_aws/tasks/vpc.yml]: I don't think it's necessary to cast to int after length.

File: roles/openshift_aws/tasks/vpc.yml [hunk 4]
```diff
@@ -1,4 +1,19 @@
 ---
+- name: query azs
+  aws_az_facts:
+  register: azs
+
+- debug:
+    msg:
```
Comment (michaelgugino) [inline roles/openshift_aws/tasks/vpc.yml]: This message should be part of failed_when, not it's own debug statement.

File: roles/openshift_aws/tasks/vpc_and_subnet_id.yml [hunk 5]
```diff
@@ -15,8 +15,14 @@
   ec2_vpc_subnet_facts:
     region: "{{ openshift_aws_region }}"
     filters:
-      "availability-zone": "{{ openshift_aws_subnet_az }}"
       vpc-id: "{{ vpcout.vpcs[0].id }}"
+      "tag:Name": "{{ item }}"
+  with_items: ['{%- if (openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute="az") | list | length) == 1 -%}
```
Comment (michaelgugino) [inline roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: This is a tad unreadable, let's add a comment here to explain what we're doing.  Also, we could probably use a vars: block to build some of these tests instead of doing it in-line with jinja.

```yaml
with_items: "{{ l_subnet_list }}"
vars:
  l_multiple_list: "{{ openshift_aws_vpc.subnets[openshift_aws_region] | map(attribute="az") | list }}"
  l_use_multiple_subnets: "{{ (l_multiple_list | length == 1) | bool }}"
  l_subnet_list: "{{ l_use_multiple_subnets | ternary([openshift_aws_subnet_az], l_multiple_list) }}"
```

**Comment by mazzystr** [comment]
8:32 sdodson: [16:36:54] ccallegar: so i think we're good to remove the default az stuff in that PR

Per irc mention from @sdodson I'm going to update this commit to remove default_az logic.

File: roles/openshift_aws/tasks/vpc_and_subnet_id.yml (hunk 5 above)
Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc_and_subnet_id.yml]: We're removing all of this per conversation regarding removing default_az.

File: roles/openshift_aws/tasks/vpc.yml (hunk 2 above)
Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc.yml]: This is necessary to pass yamllint otherwise we get...
```
$ yamllint roles/openshift_aws/tasks/vpc.yml
roles/openshift_aws/tasks/vpc.yml
  65:5      error    wrong indentation: expected 6 but found 4  (indentation)
```

File: roles/openshift_aws/tasks/vpc.yml (hunk 4 above)
Comment (mazzystr) [reply roles/openshift_aws/tasks/vpc.yml]: Ah!  TIL about "fail: msg="!

File: roles/openshift_aws/defaults/main.yml [hunk 6]
```diff
@@ -46,6 +46,21 @@ openshift_aws_ami_tags:
...
+  Name: "{{ openshift_aws_vpc_name }}"
+
+openshift_aws_vpc:
+  name: "{{ openshift_aws_vpc_name }}"
+  cidr: 172.31.0.0/16
+  subnets:
+    us-east-1:
+    - cidr: 172.31.48.0/20
+      az: "us-east-1c"
+#    - cidr: 172.31.32.0/20
```
Comment (sdodson) [inline roles/openshift_aws/defaults/main.yml]: Should these additional azs not be re-introduced now that we're considering it to be an ordered list?

Comment (mazzystr) [reply roles/openshift_aws/defaults/main.yml]: They need to be commented until multi-az is enabled since subnets/elbs/asg's now loop through this dict.  

Here is the expected behavior....
- Existing clusters / single-az: Three subnets would be deployed previously but only default_az (now element 0) would be used by ASG's and ELB's.  The other 2 subnets would be deployed across az' and continue to hand around unused.  We don't want a change to that behavior.
- New clusters / single-az: One subnet would be deployed per element 0 and be used by ASG's and ELB's.
- New multi-az clusters: Uncomment additional dictionary elements then run plays.  Three subnets would be deployed spanning AZ's but fail if more az's are requested than avail in the region.  Subnets will be utilized by ASG's and ELB's.
- Upgrade single-az to multi-az: Uncomment additional dictionary elements then run plays.   Subnets get ensured.   There won't be any changes if subnets under a count of 3 are requested as they were previously landed.  If 3+ are requested then subnets would be deployed spanning the AZ's but fail if more az's are requested than avail in the region.  ASG's and ELB's will be updated to utilize additional subnets.

Comment (sdodson) [reply roles/openshift_aws/defaults/main.yml]: Ok, thanks for the explanation.
